Index: django/core/management/commands/shell.py
===================================================================
--- django/core/management/commands/shell.py	(revision 6441)
+++ django/core/management/commands/shell.py	(working copy)
@@ -6,6 +6,8 @@
     option_list = NoArgsCommand.option_list + (
         make_option('--plain', action='store_true', dest='plain',
             help='Tells Django to use plain Python, not IPython.'),
+        make_option('--classic', action='store_true', dest='classic',
+            help='Tells Django to use IPython with the classic Python prompt.'),
     )
     help = "Runs a Python interactive interpreter. Tries to use IPython, if it's available."
 
@@ -18,6 +20,9 @@
         loaded_models = get_models()
 
         use_plain = options.get('plain', False)
+        argv = []
+        if options.get('classic', False):
+            argv.append("-cl")
 
         try:
             if use_plain:
@@ -26,7 +31,7 @@
             import IPython
             # Explicitly pass an empty list as arguments, because otherwise IPython
             # would use sys.argv from this script.
-            shell = IPython.Shell.IPShell(argv=[])
+            shell = IPython.Shell.IPShell(argv)
             shell.mainloop()
         except ImportError:
             import code
